File: game/screens/formation.lua
local focus = require("game.ui.focus")
local identity = require("render.identity")
local formations = require("data.formations")
local player_pool = require("data.players")

---@class FormationScreenContext
---@field selected string
---@field starter_ids string[]

---@class FormationScreenState
---@field viewport { w: number, h: number }
---@field selected string
---@field starter_ids string[]
---@field focus string

---@class FormationAction
---@field go "tactic"
---@field formation_id string
---@field formation string

---@class FormationScreenModule
local formation = {}

---@param ids string[]
---@return string[]
local function copy_ids(ids)
    local result = {}
    for i, id in ipairs(ids) do
        result[i] = id
    end
    return result
end

---@return table<string, PlayerData>
local function players_by_id()
    local result = {}
    for _, player in ipairs(player_pool) do
        result[player.id] = player
    end
    return result
end

---@return string[]
local function sorted_formation_ids()
    local ids = {}
    for id in pairs(formations) do
        ids[#ids + 1] = id
    end
    table.sort(ids)
    return ids
end

---@param viewport { w: number, h: number }
---@param context FormationScreenContext?
---@return FormationScreenState
function formation.new_state(viewport, context)
    local selected = context and context.selected or "2-1-1"
    assert(formations[selected], "missing formation: " .. selected)
    return {
        viewport = viewport,
        selected = selected,
        starter_ids = copy_ids(context and context.starter_ids or {}),
        focus = "formation_" .. selected,
    }
end

---@param state FormationScreenState
---@return Layout
function formation.layout(state)
    local layout = {
        {
            id = "title",
            kind = "title",
            text = "SET THE SHAPE",
            rect = { x = 0, y = 34, w = state.viewport.w, h = 34 },
            data = { align = "center" },
        },
    }
    local by_id = players_by_id()
    local markers = {}
    local lineup_names = {}
    for _, id in ipairs(state.starter_ids) do
        local player = by_id[id]
        local presentation = identity.for_player(id)
        if player then
            lineup_names[#lineup_names + 1] = player.name:upper()
        end
        markers[#markers + 1] = {
            color = presentation and presentation.palette or { 0.8, 0.9, 1 },
            shape = presentation and presentation.shape or "round",
            name = player and player.name or id,
        }
    end
    layout[#layout + 1] = {
        id = "lineup",
        kind = "label",
        text = table.concat(lineup_names, "  •  "),
        rect = { x = 80, y = 72, w = 800, h = 18 },
        data = { align = "center", tone = "muted" },
    }

    for i, id in ipairs(sorted_formation_ids()) do
        local data = formations[id]
        local y = 96 + (i - 1) * 112
        layout[#layout + 1] = {
            id = "formation_" .. id,
            kind = "button",
            text = id
                .. "  "
                .. data.name
                .. "\n+ "
                .. (data.strength or "A flexible small-sided shape.")
                .. "\n− "
                .. (data.risk or "Requires disciplined positioning."),
            selected = state.selected == id,
            focused = state.focus == "formation_" .. id,
            rect = { x = 150, y = y, w = 410, h = 94 },
            data = { align = "left" },
        }
        layout[#layout + 1] = {
            id = "preview_" .. id,
            kind = "formation_preview",
            selected = state.selected == id,
            rect = { x = 580, y = y, w = 230, h = 94 },
            data = {
                keeper = data.keeper,
                outfield = data.outfield,
                markers = markers,
            },
        }
    end
    layout[#layout + 1] = {
        id = "back",
        kind = "button",
        text = "BACK",
        focused = state.focus == "back",
        rect = { x = 254, y = 472, w = 190, h = 42 },
    }
    layout[#layout + 1] = {
        id = "next",
        kind = "button",
        text = "CHOOSE TACTIC",
        focused = state.focus == "next",
        rect = { x = 516, y = 472, w = 190, h = 42 },
    }
    return layout
end

---@param state FormationScreenState
---@param event InputEvent
---@return FormationScreenState, FormationAction|table?
function formation.update(state, event)
    local layout = formation.layout(state)
    local next = {
        viewport = state.viewport,
        selected = state.selected,
        starter_ids = copy_ids(state.starter_ids),
        focus = focus.navigate(layout, state.focus, event) or state.focus,
    }
    if event.kind == "action" and event.action == "back" then
        return next,
            {
                go = "squad",
                formation_id = next.selected,
                formation = next.selected,
            }
    end
    local id = focus.activated(layout, next.focus, event)
    if id then
        next.focus = id
        local formation_id = id:match("^formation_(.+)$") or id:match("^preview_(.+)$")
        if formation_id and formations[formation_id] then
            next.selected = formation_id
            next.focus = "formation_" .. formation_id
        elseif id == "back" then
            return next,
                {
                    go = "squad",
                    formation_id = next.selected,
                    formation = next.selected,
                }
        elseif id == "next" then
            return next,
                {
                    go = "tactic",
                    formation_id = next.selected,
                    formation = next.selected,
                }
        end
    end
    return next
end

return formation

File: ts/packages/screens/src/formation.ts
// Choose the team shape for the selected starting five. See AGENTS.md §9
// and squad.ts's header for the pure/impure seam and the content-injection
// rationale (ARCHITECTURE.md §4 rule 6): formation data, player data, and
// presentation identity come in as an explicit `FormationContentData`
// parameter instead of a module-level import.

import { focus, invariant, type Layout } from "@gc/ui";
import type { FocusEvent, FormationMarkerData } from "@gc/ui";
import type { FormationData, PlayerPresentationIdentity } from "./content.ts";

export interface FormationContentData {
  readonly formations: Readonly<Record<string, FormationData>>;
  /** The slice of player data this screen reads: just the display name. */
  readonly players: Readonly<Record<string, { readonly name: string }>>;
  readonly identities: Readonly<Record<string, PlayerPresentationIdentity>>;
}

export interface FormationScreenContext {
  readonly selected?: string;
  readonly starterIds?: readonly string[];
}

export interface FormationScreenState {
  readonly viewport: { readonly w: number; readonly h: number };
  readonly content: FormationContentData;
  readonly selected: string;
  readonly starterIds: readonly string[];
  readonly focus: string;
}

// The selection is carried as both `formationId` and `formation` on every
// transition action; `formation` is the field formation.spec.ts actually
// reads (`action.formation`), so both are kept rather than collapsed into
// one name.
export type FormationAction =
  | { readonly go: "squad"; readonly formationId: string; readonly formation: string }
  | { readonly go: "tactic"; readonly formationId: string; readonly formation: string };

function sortedFormationIds(
  formations: Readonly<Record<string, FormationData>>,
): readonly string[] {
  return Object.keys(formations).sort();
}

function newState(
  viewport: { readonly w: number; readonly h: number },
  content: FormationContentData,
  context?: FormationScreenContext,
): FormationScreenState {
  const selected = context?.selected ?? "2-1-1";
  invariant(content.formations[selected] !== undefined, `missing formation: ${selected}`);
  return {
    viewport,
    content,
    selected,
    starterIds: [...(context?.starterIds ?? [])],
    focus: `formation_${selected}`,
  };
}

function layout(state: FormationScreenState): Layout {
  const widgets: Widget[] = [
    {
      id: "title",
      kind: "title",
      text: "SET THE SHAPE",
      rect: { x: 0, y: 34, w: state.viewport.w, h: 34 },
      data: { align: "center" },
    },
  ];

  const markers: FormationMarkerData[] = [];
  const lineupNames: string[] = [];
  for (const id of state.starterIds) {
    const player = state.content.players[id];
    const presentation = state.content.identities[id];
    if (player !== undefined) {
      lineupNames.push(player.name.toUpperCase());
    }
    markers.push({
      color: presentation?.palette ?? [0.8, 0.9, 1],
      shape: presentation?.shape ?? "round",
      name: player?.name ?? id,
    });
  }
  widgets.push({
    id: "lineup",
    kind: "label",
    text: lineupNames.join("  •  "),
    rect: { x: 80, y: 72, w: 800, h: 18 },
    data: { align: "center", tone: "muted" },
  });

  sortedFormationIds(state.content.formations).forEach((id, i) => {
    const data = state.content.formations[id];
    invariant(data !== undefined, `missing formation: ${id}`);
    const y = 96 + i * 112;
    widgets.push({
      id: `formation_${id}`,
      kind: "button",
      text: `${id}  ${data.name}\n+ ${data.strength ?? "A flexible small-sided shape."}\n− ${
        data.risk ?? "Requires disciplined positioning."
      }`,
      selected: state.selected === id,
      focused: state.focus === `formation_${id}`,
      rect: { x: 150, y, w: 410, h: 94 },
      data: { align: "left" },
    });
    widgets.push({
      id: `preview_${id}`,
      kind: "formation_preview",
      selected: state.selected === id,
      rect: { x: 580, y, w: 230, h: 94 },
      data: {
        keeper: data.keeper,
        outfield: data.outfield,
        markers,
      },
    });
  });

  widgets.push({
    id: "back",
    kind: "button",
    text: "BACK",
    focused: state.focus === "back",
    rect: { x: 254, y: 472, w: 190, h: 42 },
  });
  widgets.push({
    id: "next",
    kind: "button",
    text: "CHOOSE TACTIC",
    focused: state.focus === "next",
    rect: { x: 516, y: 472, w: 190, h: 42 },
  });
  return widgets;
}

function update(
  state: FormationScreenState,
  event: FocusEvent,
): readonly [FormationScreenState, FormationAction | undefined] {
  const currentLayout = layout(state);
  let next: FormationScreenState = {
    ...state,
    starterIds: [...state.starterIds],
    focus: focus.navigate(currentLayout, state.focus, event) ?? state.focus,
  };
  if (event.kind === "action" && event.action === "back") {
    return [next, { go: "squad", formationId: next.selected, formation: next.selected }];
  }
  const id = focus.activated(currentLayout, next.focus, event);
  if (id !== null) {
    next = { ...next, focus: id };
    const formationId = (/^formation_(.+)$/.exec(id) ?? /^preview_(.+)$/.exec(id))?.[1];
    if (formationId !== undefined && state.content.formations[formationId] !== undefined) {
      next = { ...next, selected: formationId, focus: `formation_${formationId}` };
    } else if (id === "back") {
      return [next, { go: "squad", formationId: next.selected, formation: next.selected }];
    } else if (id === "next") {
      return [next, { go: "tactic", formationId: next.selected, formation: next.selected }];
    }
  }
  return [next, undefined];
}

export const formation = { newState, layout, update };

type Widget = Layout[number];
